### ISSUE TITLE
`typescript/no-unsafe-type-assertion` needs special treatment for casting from `any` and `unknown`

### ISSUE THREAD

**Issue by Atulin** [oxc-project/tsgolint#1120]
Currently, there's no way — short of pulling Zod, Typebox, or something of this kind — to deserialize JSON, for example. `const x: Type = JSON.parse(str)` and `const x = JSON.parse(str) as Type` will both cause the rule to be triggered.

The same goes for catching errors. Code like

```ts
const attempt = <T, E>(fn: () => T) : [true, T] | [false, E] => {
    try {
        const res = fn();
        return [true, res];
    } catch (e) {
        return [false, e as E];
    }
}
```

is not possible either.

If not an overarching change, then I would at least like to have a setting like

```json
{
  "options": {
    "typeAware": true
  },
  "rules": {
    "typescript/no-unsafe-type-assertion": ["warn", {
      "allowNarrowingAny": true,
      "allowNarrowingUnknown": true
    }]
  }
}
```

**Comment by camc314** [comment]
> const x: Type = JSON.parse(str) and const x = JSON.parse(str) as Type will both cause the rule to be triggered.

This is by design, the rule is designed to catch unsafe type assertions. Both of these are unsafe, as str could be anything.

> The same goes for catching errors. Code like

Same here, there are no type system features for knowing what errors a function throws, hence it is inherintly unsafe to cast from unknown to `E`.

I would suggest using an inline disable comment for these use cases to avoid loosening the type safety and correctness of your codebase.


Thanks!
